apis/controlplane/v1alpha4/nestedcomponent_types.go

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha4
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	addonv1alpha1 "sigs.k8s.io/kubebuilder-declarative-pattern/pkg/patterns/addon/pkg/apis/v1alpha1"
+)
+
+type ComponentKind string
+
+const (
+	APIServer         ComponentKind = "NestedAPIServer"
+	Etcd              ComponentKind = "NestedEtcd"
+	ControllerManager ComponentKind = "NestedControllerManager"
+)
+
+// +k8s:deepcopy-gen=false
+type NestedComponent interface {
+	client.Object
+	GetNestedComponentSpec() NestedComponentSpec
+	GetCommonStatus() addonv1alpha1.CommonStatus
+	IsComponentReady() bool
+	GetComponentKind() ComponentKind
+}
+
+func (netcd *NestedEtcd) GetNestedComponentSpec() NestedComponentSpec {
+	return netcd.Spec.NestedComponentSpec
+}
+
+func (netcd *NestedEtcd) GetCommonStatus() addonv1alpha1.CommonStatus {
+	return netcd.Status.CommonStatus
+}
+
+func (netcd *NestedEtcd) IsComponentReady() bool {
+	return netcd.Status.Phase == string(NestedEtcdReady)
+}
+
+func (netcd *NestedEtcd) GetComponentKind() ComponentKind {
+	return Etcd
+}
```

Comment:
Maybe we can move these to the specific component types files so that when you are working on a specific type you have the visibility into what might be using it? Unless we can abstract these more and make it only have these defined on the `NestedComponentSpec` since it's embedded.

Reply:
I abstract nested componenets using this interface to simplify the implementation of some common functions. E.g., 

```go
func createNestedComponentSts(..., nc capnapis.NestedComponent, ...) error {}
```.

If we move these to each type, we either need to pass `interface{}` and do type assertion in the function or we need to write `createNestedAPIServerSts` and `createNestedEtcdSts` instead, which might be verbose. Or maybe you think a little verbosity is OK. I may have read too many software engineering blogs, and "don't want to repeat my self" : )

Reply:
Totally, I'm just wondering if because those funcs are exposed on the embedded resource they should already be callable and could be more abstract. So where `IsComponentReady()` has two definitions one for etcd and one for apiserver couldn't that be one with the new reused Phase consts and just check `return obj.Status.Phase == string(NestedComponentReady)`

Reply:
aha, I see what you mean. Will remove the `NestedComponent` interface, and move these functions to the embedded struct